File: rate-limiter-service/src/main/java/com/sayam/rate_limiter_service/utility/LeakyBucket.java
package com.sayam.rate_limiter_service.utility;

import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.data.redis.core.RedisTemplate;
import org.springframework.data.redis.core.script.RedisScript;
import org.springframework.stereotype.Component;

import java.util.Collections;
import java.util.HashMap;

@Component
public class LeakyBucket {
    private final double capacity = 5;  // bucket size
    private final double leakRate = 0.1;  // tokens leaked per second

    HashMap<String, Double> keyWater = new HashMap<>();
    HashMap<String, Long> keyTimestamp = new HashMap<>();

    @Autowired
    private RedisTemplate<String, String> redisTemplate;
    @Autowired
    private RedisScript<Long> leakyBucketScript;


    public boolean allow(String apiKey) {
        String key = "ratelimit:" + apiKey;
        long now = System.currentTimeMillis();

        Long result = redisTemplate.execute(
                leakyBucketScript,
                Collections.singletonList(key),                  // KEYS[1]
                String.valueOf(capacity),                          // ARGV[1]
                String.valueOf(leakRate),                          // ARGV[2]
                String.valueOf(now)                                // ARGV[3]
        );

        return result != null && result == 1L;
    }

    private double leak(String apiKey) {
        long now = System.currentTimeMillis();

        double water = keyWater.containsKey(apiKey) ? keyWater.get(apiKey) : 0;
        double lastLeakTimestamp = keyTimestamp.containsKey(apiKey) ? keyTimestamp.get(apiKey) : now;

        double elapsedSeconds = (now - lastLeakTimestamp) / 1000.0;
        water = Math.max(0, water - elapsedSeconds * leakRate);


        keyWater.put(apiKey, water);
        keyTimestamp.put(apiKey, now);

        return water;
    }
}

File: rate-limiter-service/src/main/resources/scripts/leaky_bucket.lua
-- KEYS[1] = redis key for this apiKey's bucket
-- ARGV[1] = capacity
-- ARGV[2] = leakRate (tokens per second)
-- ARGV[3] = now (epoch millis)

local water = tonumber(redis.call("HGET", KEYS[1], "water") or "0")
local lastTs = tonumber(redis.call("HGET", KEYS[1], "ts") or ARGV[3])
local capacity = tonumber(ARGV[1])
local leakRate = tonumber(ARGV[2])
local now = tonumber(ARGV[3])

local elapsedSeconds = (now - lastTs) / 1000.0
water = math.max(0, water - elapsedSeconds * leakRate)

local allowed = 0
if water < capacity then
    water = water + 1
    allowed = 1
end

redis.call("HSET", KEYS[1], "water", tostring(water), "ts", tostring(now))
redis.call("EXPIRE", KEYS[1], 3600)

return allowed